### PR TITLE
volcano support descheduler: add framework

### PR DESCRIPTION
**Pull request by Monokaix**
part of https://github.com/volcano-sh/volcano/issues/3889

### CODE REVIEW

File: pkg/descheduler/setupplugins.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
+/*
+Copyright 2017 The Kubernetes Authors.
```
Comment (lowang-bh) [inline pkg/descheduler/setupplugins.go:2]: Does this line need to be updated?

Comment (Monokaix) [reply pkg/descheduler/setupplugins.go:2]: This is copied from k8s descheduler, it's better to keep original format?

File: pkg/descheduler/framework/profile/profile_test.go
Comment (JesseStutler) [inline pkg/descheduler/framework/profile/profile_test.go:1]: This file has no copyright header

Comment (Monokaix) [reply pkg/descheduler/framework/profile/profile_test.go:1]: done

File: pkg/descheduler/apis/componentconfig/types.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+	// Deprecated: Use clientConnection.kubeConfig instead.
+	KubeconfigFile string
+
+	// PolicyConfigFile is the filepath to the descheduler policy configuration.
+	PolicyConfigFile string
+
+	// Dry run
+	DryRun bool
+
+	// Node selectors
```
Comment (googs1025) [inline pkg/descheduler/apis/componentconfig/types.go:54]: Why is nodeselector defined as string? Can we use the fields defined in k8s directly? For example: `v1.NodeSelector`

File: pkg/descheduler/apis/componentconfig/types.go [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+	// MaxNoOfPodsToEvictPerNode restricts maximum of pods to be evicted per node.
+	MaxNoOfPodsToEvictPerNode int
+
+	// EvictLocalStoragePods allows pods using local storage to be evicted.
+	EvictLocalStoragePods bool
+
+	// IgnorePVCPods sets whether PVC pods should be allowed to be evicted
+	IgnorePVCPods bool
+
```
Comment (googs1025) [inline pkg/descheduler/apis/componentconfig/types.go:65]: If there are two fields with similar semantics, our best approach is to keep them consistent. For example: both start with Evict or both start with Igore

File: pkg/descheduler/descheduler.go [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+	if err != nil {
+		klog.Warning("Warning: Convert Kubernetes server minor version to float fail")
+	}
+
+	if math.Abs(deschedulerMinorVersionFloat-kubernetesMinorVersionFloat) > 3 {
+		klog.Warningf("Warning: Descheduler minor version %v is not supported on your version of Kubernetes %v.%v. See compatibility docs for more info: https://github.com/kubernetes-sigs/descheduler#compatibility-matrix", deschedulerMinorVersion, serverVersion.Major, serverVersion.Minor)
+	}
+}
+
+func cachedClient(
```
Comment (googs1025) [inline pkg/descheduler/descheduler.go:155]: Why do we need a fakeclient to call here?

File: pkg/descheduler/descheduler.go [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+	"sigs.k8s.io/descheduler/pkg/version"
+
+	"volcano.sh/volcano/cmd/descheduler/app/options"
+	frameworkprofile "volcano.sh/volcano/pkg/descheduler/framework/profile"
+)
+
+const podNameEnvKey string = "HOSTNAME"
+const podNamespaceEnvKey string = "POD_NAMESPACE"
+
+func Run(ctx context.Context, rs *options.DeschedulerServer) error {
```
Comment (googs1025) [inline pkg/descheduler/descheduler.go:68]: Can we separate the option validation logic into a method, such as validate or something else? This way, we don't have to write it in Run.

File: pkg/descheduler/setupplugins.go (hunk 1 above)
Comment (lowang-bh) [reply pkg/descheduler/setupplugins.go:2]: ok

**Comment by k82cn** [comment]
Why introduce an new command? It's better to use action to avoid configuration confliction between scheduler and de-scheduler.

File: pkg/descheduler/setupplugins.go (hunk 1 above)
Comment (kevin-wangzefeng) [reply pkg/descheduler/setupplugins.go:2]: We need to keep it as is, since respecting the original copyright is required by Apached 2.0 license.

**Comment by Monokaix** [comment]
> Why introduce an new command? It's better to use action to avoid configuration confliction between scheduler and de-scheduler.

That's a good question! I think there are two aspects that really matters:
1. Descheduling is a dangerous behavior and we can't sacrifice the SLO and business stability for resource utilization, so it's very important to consider the eviction condition like whether new pod can be scheduled before descheduling, and descheudler support a good mechanism to check this and it also supports a series of filter conditions like just limited nodes and namespace can do eviction, which is more responsive to users' real eviction scenarios and more versatile, while descheduling in volcano's action seems no such higher capabilities: )
2. Providing a separate descheduler component gives the user more options, e.g., if the user only wants to use real load-aware descheduling, he/she can deploy the descheduler component alone, uncoupled from the volcano scheduler itself.

File: pkg/descheduler/setupplugins.go (hunk 1 above)
Comment (Monokaix) [reply pkg/descheduler/setupplugins.go:2]: great!

File: pkg/descheduler/apis/componentconfig/types.go (hunk 2 above)
Comment (Monokaix) [reply pkg/descheduler/apis/componentconfig/types.go:54]: This is inherited from upstream descheduler and just kept from the original forma: )

File: pkg/descheduler/apis/componentconfig/types.go (hunk 3 above)
Comment (Monokaix) [reply pkg/descheduler/apis/componentconfig/types.go:65]: This is also inherited from upstream descheduler  it's better to keep same fileds.

File: pkg/descheduler/descheduler.go (hunk 4 above)
Comment (Monokaix) [reply pkg/descheduler/descheduler.go:155]: It's used in `DryRun` mode: )

File: pkg/descheduler/descheduler.go (hunk 5 above)
Comment (Monokaix) [reply pkg/descheduler/descheduler.go:68]: Same above, it's inherited from upstream descheduler.

**Comment by Monokaix** [comment]
We will Introduce new cmd in a new descheduler 
repo, see https://github.com/volcano-sh/descheduler/pull/1

**Pull request closed without merging**